libautoscoper/src/Tracker.cpp

```diff
@@ -419,61 +419,35 @@ void Tracker::optimize(int frame, int dFrame, int repeats, int opt_method, unsig
       unsigned int MAX_EPOCHS = max_iter;
       unsigned int MAX_STALL = max_stall_iter;
 
-      float positions[NUM_OF_PARTICLES*NUM_OF_DIMENSIONS];
-      float velocities[NUM_OF_PARTICLES*NUM_OF_DIMENSIONS];
-      float pBests[NUM_OF_PARTICLES*NUM_OF_DIMENSIONS];
-      float gBest[NUM_OF_DIMENSIONS];
+      // Pre-allocate particles
+      std::vector<Particle> particles(NUM_OF_PARTICLES);
 
       srand((unsigned)time(NULL));
 
-      for (int i = 0; i < NUM_OF_PARTICLES*NUM_OF_DIMENSIONS; i++)
-      {
-        if (i >= NUM_OF_DIMENSIONS) {
-          positions[i] = getRandom(START_RANGE_MIN, START_RANGE_MAX);
-        }
-        // First point will be the initial position
-        else {
-          //if (j > 1)
-          //{
-            positions[i] = (float)0.0;
-          //}
-          //else
-          //{
-          //  positions[i] = (float)0.2; // Noise for refinements
-          //}
-        }
-        pBests[i] = positions[i];
-        velocities[i] = 0;
-      }
+      // First particle is the initial position
+      particles[0] = Particle({ 0.f, 0.f, 0.f, 0.f, 0.f, 0.f });
+
+      Particle gBest = particles[0];
 
-      for (int i = 0; i < NUM_OF_DIMENSIONS; i++)
+      // ... and the other particles positions are randomly iniialized
+      for (int idx = 1; idx < NUM_OF_PARTICLES; idx++)
       {
-        gBest[i] = pBests[i];
+        particles[idx] = Particle(START_RANGE_MIN, START_RANGE_MAX);
       }
 
       clock_t cpu_begin = clock();
-
-      pso(positions, velocities, pBests, gBest, MAX_EPOCHS, MAX_STALL);
-      //cuda_pso(positions, velocities, pBests, gBest, MAX_EPOCHS);
-
+      pso(particles, gBest, MAX_EPOCHS, MAX_STALL);
```

Comment:
```suggestion
Particle gBest = pso(MAX_EPOCHS, MAX_STALL);
```

Initialization of particles and allocation of the vector can definitely all be done in the pso function.